File: api/Migrations.cs
using System.IO;
using System.Linq;
using System.Collections.Generic;
using Dapper;
using Microsoft.Extensions.Logging;
using Npgsql;

public static class Migrations
{
    public static async Task ApplyAsync(string connectionString, string sqlDirectory, ILogger logger)
    {
        if (string.IsNullOrWhiteSpace(connectionString))
        {
            logger.LogWarning("Skipping migrations because connection string is empty.");
            return;
        }

        if (!Directory.Exists(sqlDirectory))
        {
            logger.LogWarning("SQL directory '{SqlDirectory}' not found; skipping migrations.", sqlDirectory);
            return;
        }

        var files = Directory.GetFiles(sqlDirectory, "*.sql", SearchOption.TopDirectoryOnly)
            .OrderBy(f => f, StringComparer.OrdinalIgnoreCase)
            .ToList();

        if (files.Count == 0)
        {
            logger.LogInformation("No SQL migration files found in {SqlDirectory}; nothing to apply.", sqlDirectory);
            return;
        }

        await using var db = new NpgsqlConnection(connectionString);
        await db.OpenAsync();

        await using var tx = await db.BeginTransactionAsync();
        await db.ExecuteAsync(
            "create table if not exists schema_migrations(version text primary key, applied_at timestamptz not null default now());",
            transaction: tx);

        var applied = (await db.QueryAsync<string>("select version from schema_migrations", transaction: tx))
            .ToHashSet(StringComparer.OrdinalIgnoreCase);

        foreach (var file in files)
        {
            var version = Path.GetFileName(file);
            if (string.IsNullOrWhiteSpace(version))
            {
                continue;
            }

            if (applied.Contains(version))
            {
                logger.LogInformation("Migration {Version} already applied; skipping.", version);
                continue;
            }

            var sql = await File.ReadAllTextAsync(file);
            if (string.IsNullOrWhiteSpace(sql))
            {
                logger.LogWarning("Migration {Version} is empty; skipping.", version);
                continue;
            }

            logger.LogInformation("Applying migration {Version}...", version);
            await db.ExecuteAsync(sql, transaction: tx);
            await db.ExecuteAsync(
                "insert into schema_migrations(version) values(@Version);",
                new { Version = version },
                transaction: tx);
            logger.LogInformation("Migration {Version} applied.", version);
        }

        await tx.CommitAsync();
    }
}

File: apps/web-ts/backend/src/migrations.ts
import fs from "fs/promises";
import path from "path";
import { fileURLToPath } from "url";
import { query } from "./db.js";
import { env } from "./env.js";
import { logger } from "./logger.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function applyMigrations() {
  if (!env.DATABASE_URL) {
    logger.warn("DATABASE_URL not set; skipping migrations");
    return;
  }
  const sqlDir = path.resolve(__dirname, "..", "sql");
  let entries: string[] = [];
  try {
    entries = (await fs.readdir(sqlDir)).filter((f) => f.toLowerCase().endsWith(".sql"));
  } catch (err) {
    logger.warn({ err }, "SQL migrations directory not found; skipping migrations");
    return;
  }

  if (!entries.length) {
    logger.info("No SQL migrations found");
    return;
  }

  await query("create table if not exists schema_migrations(version text primary key, applied_at timestamptz not null default now())");
  const appliedRes = await query<{ version: string }>("select version from schema_migrations");
  const applied = new Set(appliedRes.rows.map((r) => r.version));

  const files = entries.sort((a, b) => a.localeCompare(b, "en"));
  for (const file of files) {
    if (applied.has(file)) {
      continue;
    }
    const fullPath = path.join(sqlDir, file);
    const sql = await fs.readFile(fullPath, "utf-8");
    if (!sql.trim()) {
      logger.warn({ file }, "Skipping empty migration");
      continue;
    }
    logger.info({ file }, "Applying migration");
    await query(sql);
    await query("insert into schema_migrations(version) values($1)", [file]);
  }
}
